docs/code_generation.md

```diff
@@ -0,0 +1,183 @@
+additional cluster info: source XML files, sdk-access methods and data types.
+There are only 2 currently in use such files:
+
+-   `src/app/zap-templates/zcl/zcl.json` is the **default** one
+-   `src/app/zap-templates/zcl/zcl-with-test-extensions.json` is in use for
+    `all-clusters-app` to show how a cluster extension is configured and has
+    minimal diffs from `zcl.json` (but it is different)
+
+### Installing zap and environment variables
+
+Matter scripts may need to invoke `zap-cli` (for code generation) or `zap` (to
+start the UI tool). For this, it needs to know where to find the commands. In
+order, the scripts will check for the following environment variables:
+
+-   if `$ZAP_DEVELOPMENT_PATH` is set, code assumes you are running zap from
+    source. Use this if you develop zap. Zap has to be bootstrapped (generally
+    `npm ci` but check zap documentation for this. Some scripts have a
+    `--run-bootstrap` command line argument to do this for you)
+
+-   if `$ZAP_INSTALL_PATH` is set, code assumes that `zap` or `zap-cli` is
+    available in the given path. This is generally an unpacked release.
+
+-   otherwise, scripts will assume `zap`/`zap-cli` is in `$PATH`
+
+### Using a UI to edit `.zap` files
+
+Generally you need to invoke zap with appropriate zcl and generate arguments.
+Most of code generation is app specific, so you generally want something of the
+form
+`--gen src/app/zap-templates/app-templates.json --zcl $ZCL_JSON_FILE $ZAP_FILE_TO_EDIT`
+
+Since this is tedious to type, the SDK provides a
+`scripts/tools/zap/run_zaptool.sh` script to automate this:
+
+```bash
+# Ensure zap is in $PATH or set $ZAP_INSTALL_PATH or $ZAP_DEVELOPMENT_PATH
+./scripts/tools/zap/run_zaptool.sh examples/lighting-app/lighting-common/lighting-app.zap
+```
+
+### Human-readable code generation inputs (`*.matter`)
+
+`.zap` files are large json files that are generally not human readable. As a
+result, the Matter SDK also keeps an equivalent `*.matter` file along side
+`.zap` files that contain the same data as `.zap` files, targeted specifically
+for matter:
+
+-   They are designed to be human readable, looking like a IDL (think protobuf
+    or android `aidl`, thrift idl etc.)
+
+-   We strive to make them contain only Matter-specific data (`.zap` files
+    contain
+-   more generic data and is designed to be ZigBee backwards compatible)
+
+Currently `.matter` file are generated from `.zap` files during the application
+specific codegen.
+
+### `*.matter` parsing and codegen
+
+`*.matter` files are both human and machine readable. Code that can process
+these files is available at `scripts/idl` and `scripts/codegen.py`. You can read
+the [Readme.md](../scripts/idl/README.md) for details of how things work.
+
+`scripts/codegen.py` can generate various outputs based on an input `*.matter`
+file.
+
+The split between `.zap` and `.matter` currently exists as an experiment of code
+generation technologies. Currently python codegen:
```

Comment:
```suggestion
generation technologies. The motivation for python code generation based on `.matter` files include:
```

Reply:
If you take this change, some rephrasing of bullets below may be required.

Reply:
I would prefer to keep old phrasing so that it shows current state rather than what we started with. When we started I think zap had more problems (e.g. it could not be properly installed as stand-alone and that was more problematic to run at compile time.